server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.opensearch.common.cache.ICacheKey;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+
+/**
+ * A class caches use to internally keep track of their stats across multiple dimensions. Not intended to be exposed outside the cache.
+ */
+public class StatsHolder implements Writeable {
+
+    // The list of permitted dimensions. Should be ordered from "outermost" to "innermost", as you would like to
+    // aggregate them in an API response.
+    private final List<String> dimensionNames;
+
+    // A map from a set of cache stats dimension values -> stats for that ordered list of dimensions.
+    private final ConcurrentMap<Key, CacheStatsResponse> statsMap;
+    CacheStatsResponse totalStats;
+
+    public StatsHolder(List<String> dimensionNames) {
+        this.dimensionNames = dimensionNames;
+        this.statsMap = new ConcurrentHashMap<>();
+        this.totalStats = new CacheStatsResponse();
+    }
+
+    public StatsHolder(StreamInput in) throws IOException {
+        this.dimensionNames = List.of(in.readStringArray());
+        Map<Key, CacheStatsResponse> readMap = in.readMap(
+            i -> new Key(List.of(i.readArray(StreamInput::readString, String[]::new))),
+            CacheStatsResponse::new
+        );
+        this.statsMap = new ConcurrentHashMap<Key, CacheStatsResponse>(readMap);
+        this.totalStats = new CacheStatsResponse(in);
+    }
```

Comment:
Why would we ever serialize a `StatsHolder`? It should be the in-memory mutable representation. We should only ever serialize the immutable snapshots of stats.

Reply:
Yes, no need to serialize this if CacheStats holds an immutable snapshot of StatsHolder.statsMap instead of StatsHolder itself.